homeassistant/components/evohome.py

```diff
@@ -0,0 +1,193 @@
+        log_level = logging.getLogger().getEffectiveLevel()
+
+        client = EvohomeClient(
+            domain_data['params'][CONF_USERNAME],
+            domain_data['params'][CONF_PASSWORD],
+            debug=False
+        )
+        # ...then restore it to what it was before instantiating the client
+        logging.getLogger().setLevel(log_level)
+
+    except RequestException as err:
+        if str(HTTP_BAD_REQUEST) in str(err):
+            # this happens when bad user credentials are supplied
+            _LOGGER.error(
+                "Failed to establish a connection with evohome web servers, "
+                "Check your username (%s), and password are correct."
+                "Unable to continue. Resolve any errors and restart HA.",
+                domain_data['params'][CONF_USERNAME]
+            )
+        else:
+            # Otherwise, it may be enough to back off and try again later.
+            raise PlatformNotReady(err)
+
+    finally:  # Redact username, password as no longer needed.
+        domain_data['params'][CONF_USERNAME] = 'REDACTED'
+        domain_data['params'][CONF_PASSWORD] = 'REDACTED'
+
+    domain_data['client'] = client
+
+    # Redact any installation data we'll never need.
+    if client.installation_info[0]['locationInfo']['locationId'] != 'REDACTED':
+        for loc in client.installation_info:
+            loc['locationInfo']['locationId'] = 'REDACTED'
+            loc['locationInfo']['streetAddress'] = 'REDACTED'
+            loc['locationInfo']['city'] = 'REDACTED'
+            loc['locationInfo']['locationOwner'] = 'REDACTED'
+            loc[GWS][0]['gatewayInfo'] = 'REDACTED'
+
+    # Pull down the installation configuration...
+    loc_idx = domain_data['params'][CONF_LOCATION_IDX]
+
+    try:
+        domain_data['config'] = client.installation_info[loc_idx]
+
+    # IndexError: configured location index is outside the range
+    except IndexError:
+        _LOGGER.warning(
+            "setup(): Config parameter, '%s'= %s , is out of range (0-%s)",
+            CONF_LOCATION_IDX,
+            loc_idx,
+            len(client.installation_info) - 1
+        )
+
+        raise
+
+    domain_data['status'] = {}
+
+    if _LOGGER.isEnabledFor(logging.DEBUG):
+        _LOGGER.debug(
+            "setup(): Location/TCS (temp. control system) used is: %s [%s]",
+            domain_data['config'][GWS][0][TCS][0]['systemId'],
+            domain_data['config']['locationInfo']['name']
+        )
+        # Some platform data needs further redaction before being logged.
+        tmp = dict(domain_data['config'])
+        tmp['locationInfo']['postcode'] = 'REDACTED'
+
+        _LOGGER.debug("setup(): domain_data['config']: %s", tmp)
+
+    # We have the platofrom configuration, but no state as yet, so...
+    def _first_update(event):
+        """Let the controller know it can obtain it's first update."""
+    # Send a message to the hub to do its first update()
```

Comment:
Remove this.

Reply:
Tidied up.  Removed triple-quotes as private method.